### PR TITLE
Add Zalo provider

### PR DESCRIPTION
**Pull request by hoangitk**
Zalo is a free message and call application on mobile. And most popular in Vietnam.

Docs: https://developers.zalo.me/docs/api/social-api-4

### CODE REVIEW

File: src/AspNet.Security.OAuth.Zalo/ZaloAuthenticationDefaults.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
+﻿using Microsoft.AspNetCore.Authentication;
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Zalo/ZaloAuthenticationDefaults.cs]: Please add the header to these source files. Example:

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/542ac858ea1d1be1ca6fab2dacb37d4fd309248f/src/AspNet.Security.OAuth.Amazon/AmazonAuthenticationDefaults.cs#L1-L5

File: src/AspNet.Security.OAuth.Zalo/ZaloAuthenticationDefaults.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+        /// <summary>
+        /// Default value for <see cref="RemoteAuthenticationOptions.CallbackPath"/>.
+        /// </summary>
+        public const string CallbackPath = "/signin-zalo";
+
+        /// <summary>
+        /// Default value for <see cref="OAuthOptions.AuthorizationEndpoint"/>.
+        /// </summary>
+        public const string AuthorizationEndpoint = "https://oauth.zaloapp.com/v3/auth";
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Zalo/ZaloAuthenticationDefaults.cs:37]: Just to check, the documentation you linked to was to a version 4 API, but these endpoints are two version 3 and one version 2 - is that definitely correct?

File: src/AspNet.Security.OAuth.Zalo/ZaloAuthenticationExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        public static AuthenticationBuilder AddZalo(
+            [NotNull] this AuthenticationBuilder builder,
+            [NotNull] Action<ZaloAuthenticationOptions> configuration)
+        {
+            return builder.AddZalo(
+                ZaloAuthenticationDefaults.AuthenticationScheme,
+                ZaloAuthenticationDefaults.DisplayName,
+                configuration);
+        }
+
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Zalo/ZaloAuthenticationExtensions.cs:57]: The other providers provider a 4th extension method. For consistency can you add the equivalent please?

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/542ac858ea1d1be1ca6fab2dacb37d4fd309248f/src/AspNet.Security.OAuth.Amazon/AmazonAuthenticationExtensions.cs#L44-L57

File: src/AspNet.Security.OAuth.Zalo/ZaloAuthenticationHandler.cs [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+            context.RunClaimActions(payload);
+
+            await Options.Events.CreatingTicket(context);
+            return new AuthenticationTicket(context.Principal, context.Properties, Scheme.Name);
+        }
+
+        protected override string BuildChallengeUrl(AuthenticationProperties properties, string redirectUri)
+        {
+            var address = base.BuildChallengeUrl(properties, redirectUri);
+            return QueryHelpers.AddQueryString(address, new Dictionary<string, string>
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Zalo/ZaloAuthenticationHandler.cs]: You can use the `QueryHelpers.AddQueryString(string, string, string)` overload here instead.

File: src/AspNet.Security.OAuth.Zalo/ZaloAuthenticationOptions.cs [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+            TokenEndpoint = ZaloAuthenticationDefaults.TokenEndpoint;
+            UserInformationEndpoint = ZaloAuthenticationDefaults.UserInformationEndpoint;
+
+            ClaimActions.MapJsonKey(ClaimTypes.NameIdentifier, "id");
+            ClaimActions.MapJsonKey(ClaimTypes.Name, "name");
+            ClaimActions.MapJsonKey(ClaimTypes.Gender, "gender");
+            ClaimActions.MapCustomJson(ClaimTypes.DateOfBirth, user =>
+            {
+                return DateTime.TryParseExact(user.Value<string>("birthday"), "dd/MM/yyyy", CultureInfo.InvariantCulture, DateTimeStyles.None, out var dateOfBirth)
+                    ? dateOfBirth.ToString("yyyy/MM/dd")
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Zalo/ZaloAuthenticationOptions.cs]: Use `dateOfBirth.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture)`. An ISO format is more interoperable.

File: test/AspNet.Security.OAuth.Providers.Tests/Zalo/ZaloTests.cs [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        public override string DefaultScheme => ZaloAuthenticationDefaults.AuthenticationScheme;
+
+        protected internal override void RegisterAuthentication(AuthenticationBuilder builder)
+        {
+            builder.AddZalo(options => ConfigureDefaults(builder, options));
+        }
+
+        [Theory]
+        [InlineData(ClaimTypes.NameIdentifier, "my-id")]
+        [InlineData(ClaimTypes.Name, "HoangITK")]
```
Comment (martincostello) [inline test/AspNet.Security.OAuth.Providers.Tests/Zalo/ZaloTests.cs:26]: You should add a test case for the `DateOfBirth` and `Gender` claims as you're adding code to deal with them.

**Comment by martincostello** [comment]
Can you also add the provider to the README please? There's a template commented-out at the end of the file.

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/dev/README.md#providers

File: src/AspNet.Security.OAuth.Zalo/ZaloAuthenticationDefaults.cs (hunk 2 above)
Comment (hoangitk) [reply src/AspNet.Security.OAuth.Zalo/ZaloAuthenticationDefaults.cs:37]: + Yes, it is correct. 
+ social-api-4: 4 is id of cms page.

File: src/AspNet.Security.OAuth.Zalo/ZaloAuthenticationExtensions.cs [hunk 7]
```diff
@@ -24,16 +30,29 @@ public static class ZaloAuthenticationExtensions
...
-            return builder.AddZalo(
-                ZaloAuthenticationDefaults.AuthenticationScheme,
-                ZaloAuthenticationDefaults.DisplayName,
-                configuration);
+            return builder.AddZalo(ZaloAuthenticationDefaults.AuthenticationScheme, ZaloAuthenticationDefaults.DisplayName, configuration);
+        }
+
+        /// <summary>
+        /// Adds <see cref="ZaloAuthenticationHandler"/> to the specified
+        /// <see cref="AuthenticationBuilder"/>, which enables Amazon authentication capabilities.
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Zalo/ZaloAuthenticationExtensions.cs]: _Amazon_ -> _Zalo_

File: src/AspNet.Security.OAuth.Zalo/ZaloAuthenticationExtensions.cs [hunk 8]
```diff
@@ -24,16 +30,29 @@ public static class ZaloAuthenticationExtensions
...
+        /// </summary>
+        /// <param name="builder">The authentication builder.</param>
+        /// <param name="scheme">The authentication scheme associated with this instance.</param>
+        /// <param name="configuration">The delegate used to configure the Zalo options.</param>
+        /// <returns>The <see cref="AuthenticationBuilder"/>.</returns>
+        public static AuthenticationBuilder AddAmazon(
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Zalo/ZaloAuthenticationExtensions.cs]: `AddZalo()`

Comment (hoangitk) [reply src/AspNet.Security.OAuth.Zalo/ZaloAuthenticationExtensions.cs]: 😅

**Review by martincostello — APPROVED** [review]

**Comment by martincostello** [comment]
Great work @hoangitk, thank you for your contribution! 🎈

**Comment by hoangitk** [comment]
Thank for your review!

**Comment by martincostello** [comment]
This provider has now been deployed to [NuGet.org](https://www.nuget.org/packages/AspNet.Security.OAuth.Zalo/2.1.0) - thanks again for your contribution @hoangitk! 🎉

**Pull request merged**